File: solutions/cpp/luhn/1/luhn.cpp
#include "luhn.h"
#include <vector>

namespace luhn {
    bool valid(const std::string& ccnumber) {
        auto start = ccnumber.begin();
        auto end = ccnumber.end();
        while (std::isspace(*start)) {
            ++start;
        }
        while (std::isspace(*(end - 1))) {
            --end;
        }

        if (end - start < 2) {
            return false;
        }

        std::vector<int> digits;
        while (start != end) {
            char c = *start;
            ++start;
            if (std::isspace(c)) {
                continue;
            } else if (std::isdigit(c)) {
                digits.emplace_back(c - '0');
            } else {
                return false;
            }
        }
        
        int sum = 0; 
        for (int i = digits.size() - 1, count = 0; i >= 0; --i, ++count) {
            int digit = digits[i];
            if (count % 2 == 1) {
                digit *= 2;
                if (digit > 9) {
                    digit -= 9;
                }
            }
            sum += digit;
        }
        return sum % 10 == 0;
    }
}

File: solutions/clojure/luhn/1/src/luhn.clj
(ns luhn)

(defn to-digit [c]
  (let [zero (int \0)]
    (- (int c) zero)))

(defn double-digit [digit]
  (let [result (* 2 digit)]
    (if (> result 9)
      (- result 9)
      result)))

(defn divisible-by-10 [n]
  (= 0 (rem n 10)))

(defn valid? [s]
  (let [digits (->> s
                (filter (partial not= \space))
                (map to-digit))]
    (if (> 2 (count digits))
      false
      (->> digits
           reverse
           (partition 2)
           (map #(list (first %) (double-digit (second %))))
           flatten
           (apply +)
           divisible-by-10))))
